internal/xds/bootstrap/bootstrap.go

```
@@ -227,6 +257,24 @@ func (sc *ServerConfig) DialOptions() []grpc.DialOption {
 	return dopts
 }
 
+// DialOptionsWithCallCredsForTransport returns dial options with call
+// credentials but only if they are compatible with the specified transport
+// credentials type. For example, JWT call credentials require secure transport
+// and are skipped for insecure.
+func (sc *ServerConfig) DialOptionsWithCallCredsForTransport(transportCredsType string, transportCreds credentials.TransportCredentials) []grpc.DialOption {
+	dopts := sc.DialOptions()
+	isInsecureTransport := transportCredsType == "insecure" ||
+		(transportCreds != nil && transportCreds.Info().SecurityProtocol == "insecure")
+
+	for _, callCred := range sc.selectedCallCreds {
+		if isInsecureTransport && callCred.RequireTransportSecurity() {
+			continue
+		}
+		dopts = append(dopts, grpc.WithPerRPCCredentials(callCred))
+	}
+	return dopts
+}
```

The gRFC states the following in the documentation for the new field `call_creds` in the bootstrap configuration.

```
  // NEW FIELD!
  // List of call creds.  Optional.  Client will apply all call creds
  // types that it supports but will ignore any types that it does not
  // support.
```

My reading of this is that we *should* apply all supported call creds unconditionally. And if it so happens that an insecure transport is configured, then we can let the RPC fail and that will result in a more meaningful error to the user (something like cannot send secure call creds on an insecure connection). I will check what the other language implementations plan to do and will get back to you on this.

Also, I'm not very thrilled about the way `isInsecureTransport` is determined. This seems a bit hacky to me and would require changes if another insecure channel credentials implementation is added in the future which has a name other than `insecure`.

---

I checked with other grpc language leads and they seem to agree that we don't want to be doing this check. We should let the check happen at RPC time. The one thing that we need to verify though is that a meaningful error is propagated to the user when a failure happens here. Thanks.

---

WRT the meaningful error, I _think_ what will happen is https://github.com/grpc/grpc-go/blob/master/credentials/jwt/token_file_call_creds.go#L78 will be called and you'd get `cannot send secure credentials on an insecure connection: requires SecurityLevel PrivacyAndIntegrity; connection has NoSecurity`

---

I filed https://github.com/grpc/grpc-go/issues/8635 to keep track of this. Thanks.